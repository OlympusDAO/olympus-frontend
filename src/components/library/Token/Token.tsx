import { SvgIcon, SvgIconProps } from "@mui/material";
import { styled } from "@mui/material/styles";
import { FC } from "react";
import tokenPath from "src/components/library/Token/tokensLib";

const PREFIX = "Token";

const classes = {
  root: `${PREFIX}-root`,
};

const StyledSvgIcon = styled(SvgIcon)(() => ({
  [`&.${classes.root}`]: {
    display: "flex",
    justifyContent: "space-between",
    flexDirection: "row",
    margin: "12px 0px",
  },
}));

export interface OHMTokenProps extends SvgIconProps {
  name: keyof typeof tokenPath;
  viewBox?: string;
  fontSize?: SvgIconProps["fontSize"];
}
/**
 * Primary Token Component for UI. This component displays a SVG Token Asset from the OlympusDAO Token Library.
 *
 * ## Updating Component Library with a New Token Asset
 * You may find that a specific token asset is not available inside of the component library package. To add a new token to component-library follow the following steps.
 * 
 * #### Step 1
 * Ensure your SVG file is optimized using svgo. Run ```npx svgo <yourasset.svg> ```
 * 
 * #### Step 2
 * We've tried to make sure default viewports, viewboxes and overall heights/widths are aligned to a standard 32x32 as a base default. 
 *  https://boxy-svg.com/ is a great resource to verify that the default width/height of the SVG is set properly and that the viewbox is set to 32x32. Once verified/modified, export the SVG and move to the next step.
 * 
 * #### Step 3
 * Convert your SVG to a react component by running ```npx @svgr/cli <YourProperlySized.svg>```. SVGR will automatically normalize your SVG and rename attributes to play nice with react. 
 * You'll see a react component output to the command line. Copy the output BETWEEN the starting and closing ```<svg></svg>``` tags. Exclude the starting/ending ```<svg>``` tags. 
 *
 *  #### Step 4
 * Add a new Key:Value to tokensLib.tsx with Key being ALL CAPS for a Token Symbol or all lowercase for a network/other type of token. 
 * Example
 * 
 * ```jsx
 *   AVAX: (
 *   <>
 *     <path
 *       style={{
 *         fill: "#fff",
 *         strokeWidth: 0.02130493,
 *       }}
 *       d="M6.104 5.475h19.771v17.981H6.104z"
 *     />
 *     <path
 *       d="M32 16c0 8.837-7.163 16-16 16S0 24.837 0 16 7.163 0 16 0s16 7.163 16 16Zm-20.534 6.367H8.361c-.653 0-.975 0-1.171-.126a.79.79 0 0 1-.358-.617c-.012-.232.15-.515.472-1.08L14.97 7.028c.326-.574.49-.86.7-.967a.791.791 0 0 1 .715 0c.208.106.373.393.7.967L18.66 9.78l.008.014c.353.615.531.927.61 1.255.086.358.086.735 0 1.093-.08.33-.256.644-.614 1.27l-4.027 7.119-.01.018c-.355.62-.535.935-.784 1.172-.271.26-.597.448-.955.555-.326.09-.692.09-1.423.09zm7.842 0h4.449c.656 0 .987 0 1.183-.13a.787.787 0 0 0 .358-.62c.011-.225-.146-.497-.455-1.03l-.033-.055-2.228-3.813-.026-.043c-.313-.53-.471-.797-.674-.9a.783.783 0 0 0-.711 0c-.205.106-.37.385-.696.947l-2.22 3.813-.009.013c-.325.56-.487.841-.476 1.071a.796.796 0 0 0 .358.622c.193.125.523.125 1.18.125z"
 *       style={{
 *         clipRule: "evenodd",
 *         fill: "#e84142",
 *         fillRule: "evenodd",
 *         strokeWidth: 0.02130493,
 *       }}
 *     />
 *   </>
 * ),
  ```
 *  
 * #### Step 5
 * Submit a PR to the component-library project, to be published in the next npm package release. 
 */

const Token: FC<OHMTokenProps> = ({ name, viewBox = "0 0 32 32", fontSize = "large", ...props }) => {
  return (
    <StyledSvgIcon viewBox={viewBox} fontSize={fontSize} {...props}>
      {tokenPath[name]}
    </StyledSvgIcon>
  );
};

export default Token;
export type TokenName = keyof typeof tokenPath;
