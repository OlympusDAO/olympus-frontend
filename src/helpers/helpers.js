export const BOND_AMOUNT = 10;
// Sometimes we need to round float values because bigint type does not exist (yet) in javascript
export function round(val) {
  var m = Math.pow(10, 10);
  return Math.round(parseFloat(val) * m) / m;
}
