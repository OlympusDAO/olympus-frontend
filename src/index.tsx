// import { ApolloClient, ApolloProvider, InMemoryCache } from "@apollo/client";

// import React from "react";
import ReactDOM from "react-dom";
import Root from "./Root";

// const subgraphUri = "http://localhost:8000/subgraphs/name/scaffold-eth/your-contract";

// const client = new ApolloClient({
//   uri: subgraphUri,
//   cache: new InMemoryCache(),
// });

ReactDOM.render(<Root />, document.getElementById("root"));
