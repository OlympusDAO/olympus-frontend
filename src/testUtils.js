import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "@material-ui/core/styles";
import CssBaseline from "@material-ui/core/CssBaseline";
import { light as lightTheme } from "./themes/light.js";
import { Web3ContextProvider } from "./hooks/web3Context";
import store from "./store.ts";
import { createMemoryHistory } from "history";
import App from "./App";
import { I18nProvider } from "@lingui/react";
import { i18n } from "@lingui/core";

const AllTheProviders = ({ children }) => {
  return (
    <Web3ContextProvider>
      <Provider store={store}>
        <BrowserRouter basename={"/"}>
          <ThemeProvider theme={lightTheme}>
            <CssBaseline />
            {children}
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    </Web3ContextProvider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

const renderRoute = function (route) {
  const history = createMemoryHistory();
  history.push(route);
  render(
    <Web3ContextProvider>
      <Provider store={store}>
        <I18nProvider i18n={i18n}>
          <BrowserRouter basename={"/#"}>
            <App />
          </BrowserRouter>
        </I18nProvider>
      </Provider>
    </Web3ContextProvider>,
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render, renderRoute };
