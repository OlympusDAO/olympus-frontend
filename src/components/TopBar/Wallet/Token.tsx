import { t } from "@lingui/macro";
import {
  Accordion as MuiAccordion,
  AccordionDetails,
  AccordionSummary as MuiAccordionSummary,
  Box,
  Button,
  Typography,
  useTheme,
  withStyles,
} from "@material-ui/core";
import { Skeleton } from "@material-ui/lab";
import { Icon, OHMTokenProps, Token as TokenSVG } from "@olympusdao/component-library";
import { ChangeEvent, useState } from "react";
import { useQuery } from "react-query";
import { addresses, NETWORKS } from "src/constants";
import { NetworkId } from "src/constants";
import { formatCurrency } from "src/helpers";
import { useAppSelector } from "src/hooks";
import { useWeb3Context } from "src/hooks/web3Context";
import { fetchCrossChainBalances } from "src/lib/fetchBalances";

import { trackGAEvent, trackSegmentEvent } from "../../../helpers/analytics";

const Accordion = withStyles({
  root: {
    backgroundColor: "inherit",
    backdropFilter: "none",
    "-webkit-backdrop-filter": "none",
    boxShadow: "none",
    "&:before": {
      display: "none",
    },
    "&$expanded": {
      margin: 0,
    },
  },
  expanded: {},
})(MuiAccordion);

const AccordionSummary = withStyles(theme => ({
  root: {
    minHeight: "36px",
    height: "36px",
    padding: theme.spacing(0),
    "&$expanded": {
      padding: theme.spacing(0),
      minHeight: "36px",
    },
  },
  content: {
    margin: 0,
    "&$expanded": {
      margin: 0,
    },
  },
  expanded: {},
}))(MuiAccordionSummary);

export interface IToken {
  symbol: string;
  address: string;
  decimals: number;
  icon: OHMTokenProps["name"];
  balance: string;
  price: number;
  crossChainBalances?: { balances: Record<NetworkId, string>; isLoading: boolean };
  vaultBalances?: { [vaultName: string]: string };
  totalBalance: string;
}

const addTokenToWallet = async (token: IToken, userAddress: string) => {
  if (!window.ethereum) return;
  const host = window.location.origin;
  try {
    await window.ethereum.request({
      method: "wallet_watchAsset",
      params: {
        type: "ERC20",
        options: {
          address: token.address,
          symbol: token.symbol,
          decimals: token.decimals,
          image: `${host}/${token.icon}`,
        },
      },
    });
    trackSegmentEvent({
      address: userAddress,
      type: "Add Token",
      tokenName: token.symbol,
    });
    trackGAEvent({
      category: "Token",
      action: "Add Token",
      label: token.symbol ?? "unknown",
    });
  } catch (error) {
    console.log(error);
  }
};

interface TokenProps extends IToken {
  expanded: boolean;
  onChangeExpanded: (event: ChangeEvent<any>, isExpanded: boolean) => void;
  onAddTokenToWallet: () => void;
  decimals: number;
}

const BalanceValue = ({
  balance,
  balanceValueUSD,
  isLoading = false,
  sigFigs,
}: {
  balance: string;
  balanceValueUSD: number;
  isLoading?: boolean;
  sigFigs: number;
}) => (
  <Box sx={{ textAlign: "right", display: "flex", flexDirection: "column", justifyContent: "space-between" }}>
    <Typography variant="body2" style={{ fontWeight: 600 }}>
      {!isLoading ? balance.substring(0, sigFigs) : <Skeleton variant="text" width={50} />}
    </Typography>
    <Typography variant="body2" color="textSecondary">
      {!isLoading ? (
        formatCurrency(balanceValueUSD === NaN ? 0 : balanceValueUSD, 2)
      ) : (
        <Skeleton variant="text" width={50} />
      )}
    </Typography>
  </Box>
);

const TokenBalance = ({
  balanceLabel,
  balance,
  balanceValueUSD,
  sigFigs,
}: {
  balanceLabel: string;
  balance: string;
  balanceValueUSD: number;
  sigFigs: number;
}) => (
  <Box display="flex" flexDirection="row" justifyContent="space-between" key={balanceLabel}>
    <Typography color="textSecondary">{balanceLabel}</Typography>
    <Typography color="textSecondary">
      <BalanceValue balance={balance} sigFigs={sigFigs} balanceValueUSD={balanceValueUSD} />
    </Typography>
  </Box>
);

export const Token = ({
  symbol,
  decimals,
  icon,
  price = 0,
  crossChainBalances,
  vaultBalances,
  totalBalance,
  onAddTokenToWallet,
  expanded,
  onChangeExpanded,
}: TokenProps) => {
  const theme = useTheme();
  const isLoading = useAppSelector(s => s.account.loading || s.app.loadingMarketPrice || s.app.loading);
  const balanceValue = parseFloat(totalBalance) * price;

  // cleanedDecimals provides up to 7 sigFigs on an 18 decimal token (gOHM) & 5 sigFigs on 9 decimal Token
  const sigFigs = decimals === 18 ? 7 : 5;

  return (
    <Accordion expanded={expanded} onChange={onChangeExpanded}>
      <AccordionSummary expandIcon={<Icon name="more" color="disabled" />}>
        <Box sx={{ display: "flex", alignItems: "center", width: "100%" }}>
          <TokenSVG name={icon} style={{ fontSize: 28, marginRight: theme.spacing(1) }} />
          <Typography>{symbol}</Typography>
        </Box>
        <BalanceValue
          balance={totalBalance}
          sigFigs={sigFigs}
          balanceValueUSD={balanceValue}
          isLoading={isLoading || crossChainBalances?.isLoading}
        />
      </AccordionSummary>
      <AccordionDetails style={{ margin: "auto", padding: theme.spacing(1, 0) }}>
        <Box
          sx={{ display: "flex", flexDirection: "column", flex: 1, mx: "32px", justifyContent: "center" }}
          style={{ gap: theme.spacing(1) }}
        >
          {!!crossChainBalances?.balances &&
            Object.entries(crossChainBalances.balances).map(
              ([networkId, balance]) =>
                parseFloat(balance) > 0.01 && (
                  <TokenBalance
                    balanceLabel={`${NETWORKS[networkId as any].chainName}:`}
                    balance={balance}
                    balanceValueUSD={parseFloat(balance) * price}
                    sigFigs={sigFigs}
                  />
                ),
            )}
          {!!vaultBalances &&
            Object.entries(vaultBalances).map(
              ([vaultName, balance]) =>
                parseFloat(balance) > 0.01 && (
                  <TokenBalance
                    balanceLabel={`${vaultName}:`}
                    balance={balance}
                    balanceValueUSD={parseFloat(balance) * price}
                    sigFigs={sigFigs}
                  />
                ),
            )}
          <Box className="ohm-pairs" style={{ width: "100%" }}>
            <Button variant="contained" color="secondary" fullWidth onClick={onAddTokenToWallet}>
              <Typography>{t`Add to Wallet`}</Typography>
            </Button>
          </Box>
        </Box>
      </AccordionDetails>
    </Accordion>
  );
};

export const MigrateToken = ({ symbol, icon, balance = "0.0", price = 0 }: IToken) => {
  const theme = useTheme();
  const balanceValue = parseFloat(balance) * price;
  return (
    <Box sx={{ display: "flex", flexDirection: "row" }}>
      <Box sx={{ display: "flex", alignItems: "center", width: "100%" }}>
        <TokenSVG name={icon} style={{ fontSize: 28, marginRight: theme.spacing(1) }} />
        <Typography>{symbol}</Typography>
      </Box>
      {/* <Button variant="contained" color="primary" size="small" onClick={() => true}>
        Migrate v2
      </Button> */}
      <Box
        sx={{
          mx: theme.spacing(0.5),
          textAlign: "right",
          display: "flex",
          flexDirection: "column",
          justifyContent: "space-between",
        }}
      >
        <Typography variant="body2" style={{ fontWeight: 600 }}>
          {balance.substring(0, 5)}
        </Typography>
        <Typography variant="body2" color="textSecondary">
          {formatCurrency(balanceValue, 2)}
        </Typography>
      </Box>
    </Box>
  );
};

const sumObjValues = (obj: Record<string, string> = {}) =>
  Object.values(obj).reduce((sum, b = "0.0") => sum + (parseFloat(b) || 0), 0);

export const useWallet = (
  userAddress: string,
  chainId: NetworkId,
  providerInitialized: boolean,
): Record<string, IToken> => {
  // default to mainnet while not initialized
  const networkId = providerInitialized ? chainId : NetworkId.MAINNET;

  const connectedChainBalances = useAppSelector(s => s.account.balances);
  const ohmPrice = useAppSelector(s => s.app.marketPrice);
  const currentIndex = useAppSelector(s => s.app.currentIndex);

  const { gohm, wsohm, isLoading } = useCrossChainBalances(userAddress);

  const tokens = {
    ohmV1: {
      symbol: "OHM V1",
      address: addresses[networkId].OHM_ADDRESS,
      balance: connectedChainBalances.ohmV1,
      price: ohmPrice || 0,
      icon: "OHM",
      decimals: 9,
    },
    sohmV1: {
      symbol: "sOHM V1",
      address: addresses[networkId].SOHM_ADDRESS,
      balance: connectedChainBalances.sohmV1,
      price: ohmPrice || 0,
      icon: "sOHM",
      decimals: 9,
    },
    ohm: {
      symbol: "OHM",
      address: addresses[networkId].OHM_V2,
      balance: connectedChainBalances.ohm,
      price: ohmPrice || 0,
      icon: "OHM",
      decimals: 9,
    },
    sohm: {
      symbol: "sOHM",
      address: addresses[networkId].SOHM_V2,
      balance: connectedChainBalances.sohm,
      price: ohmPrice || 0,
      vaultBalances: {
        "Fuse Olympus Pool Party": connectedChainBalances.fsohm,
      },
      icon: "sOHM",
      decimals: 9,
    },
    wsohm: {
      symbol: "wsOHM",
      address: addresses[networkId].WSOHM_ADDRESS,
      balance: connectedChainBalances.wsohm,
      price: (ohmPrice || 0) * Number(currentIndex || 0),
      crossChainBalances: { balances: wsohm, isLoading },
      icon: "wsOHM",
      decimals: 18,
    },
    pool: {
      symbol: "33T",
      address: addresses[networkId].PT_TOKEN_ADDRESS,
      balance: connectedChainBalances.pool,
      price: ohmPrice || 0,
      icon: "33T",
      decimals: 9,
    },
    gohm: {
      symbol: "gOHM",
      address: addresses[networkId].GOHM_ADDRESS,
      balance: connectedChainBalances.gohm,
      price: (ohmPrice || 0) * Number(currentIndex || 0),
      crossChainBalances: { balances: gohm, isLoading },
      vaultBalances: {
        "gOHM on Tokemak": connectedChainBalances.gOhmOnTokemak,
        "Fuse Olympus Pool Party": connectedChainBalances.fgohm,
      },
      icon: "wsOHM",
      decimals: 18,
    },
  } as Record<string, Omit<IToken, "totalBalance">>;

  return Object.entries(tokens).reduce((wallet, [key, token]) => {
    const crossChainBalances = sumObjValues(token.crossChainBalances?.balances);
    const vaultBalances = sumObjValues(token.vaultBalances);
    const balance = crossChainBalances || parseFloat(token.balance) || 0;
    return {
      ...wallet,
      [key]: {
        ...token,
        totalBalance: (balance + vaultBalances).toString(),
      } as IToken,
    };
  }, {});
};

export const useCrossChainBalances = (address: string) => {
  const { isLoading, data } = useQuery(["crossChainBalances", address], () => fetchCrossChainBalances(address), {
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
  });
  return { isLoading, ...data };
};

export const Tokens = () => {
  const { address: userAddress, networkId, providerInitialized } = useWeb3Context();
  const tokens = useWallet(userAddress, networkId, providerInitialized);
  const isLoading = useAppSelector(s => s.account.loading || s.app.loadingMarketPrice || s.app.loading);
  const [expanded, setExpanded] = useState<string | null>(null);

  const v1Tokens = [tokens.ohmV1, tokens.sohmV1];
  const alwaysShowTokens = [tokens.ohm, tokens.sohm, tokens.gohm];
  const onlyShowWhenBalanceTokens = [tokens.wsohm, tokens.pool];

  const tokenProps = (token: IToken) => ({
    ...token,
    expanded: expanded === token.symbol,
    onChangeExpanded: (e: any, isExpanded: boolean) => setExpanded(isExpanded ? token.symbol : null),
    onAddTokenToWallet: () => addTokenToWallet(token, userAddress),
  });

  return (
    <>
      {alwaysShowTokens.map(token => (
        <Token key={token.symbol} {...tokenProps(token)} />
      ))}
      {!isLoading &&
        onlyShowWhenBalanceTokens.map(
          token => parseFloat(token.totalBalance) > 0.01 && <Token key={token.symbol} {...tokenProps(token)} />,
        )}
      {!isLoading &&
        v1Tokens.map(token => parseFloat(token.totalBalance) > 0.01 && <MigrateToken {...token} key={token.symbol} />)}
    </>
  );
};
