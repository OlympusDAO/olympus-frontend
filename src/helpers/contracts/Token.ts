import { OHMTokenStackProps } from "@olympusdao/component-library";
import { AddressMap } from "src/constants/addresses";
import { Contract, ContractConfig, Factory } from "src/helpers/contracts/Contract";
import { DecimalBigNumber } from "src/helpers/DecimalBigNumber/DecimalBigNumber";
import { NetworkId } from "src/networkDetails";

import { getCoingeckoPrice } from "../pricing/getCoingeckoPrice";
import { assert } from "../types/assert";

export interface TokenConfig<TFactory extends Factory = Factory, TAddressMap extends AddressMap = AddressMap>
  extends ContractConfig<TFactory, TAddressMap> {
  decimals: number;
  purchaseUrl: string;
  icons: NonNullable<OHMTokenStackProps["tokens"]>;
  customPricingFunc?: (networkId: keyof TAddressMap) => Promise<DecimalBigNumber>;
}

export class Token<TFactory extends Factory = Factory, TAddressMap extends AddressMap = AddressMap> extends Contract<
  TFactory,
  TAddressMap
> {
  /**
   * An array of icons for this token
   *
   * Commonly, there is only one icon. For LP tokens, there are two.
   */
  icons: TokenConfig<TFactory, TAddressMap>["icons"];

  /**
   * The number of decimal used by the tokens ERC20 contract.
   */
  decimals: TokenConfig<TFactory, TAddressMap>["decimals"];

  /**
   * A url where this token can be purchased
   */
  purchaseUrl: TokenConfig<TFactory, TAddressMap>["purchaseUrl"];

  /**
   * Function that returns the price of this token in USD
   */
  customPricingFunc?: TokenConfig<TFactory, TAddressMap>["customPricingFunc"];

  constructor(config: TokenConfig<TFactory, TAddressMap>) {
    super({ name: config.name, factory: config.factory, addresses: config.addresses });

    this.icons = config.icons;
    this.decimals = config.decimals;
    this.purchaseUrl = config.purchaseUrl;
    this.customPricingFunc = config.customPricingFunc;
  }

  async getPrice(networkId: keyof TAddressMap) {
    if (this.customPricingFunc) return this.customPricingFunc(networkId);

    const address = this.addresses[networkId];
    assert(address, `Address should exist for token: ${this.name} on network: ${networkId.toString()}`);

    // Default to coingecko
    return getCoingeckoPrice(networkId as NetworkId, address as unknown as string);
  }
}
