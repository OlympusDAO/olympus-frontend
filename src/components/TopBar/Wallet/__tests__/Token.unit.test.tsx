import { screen } from "@testing-library/react";
import { getByTestId } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { formatCurrency } from "src/helpers";

import { render } from "../../../../testUtils";
import { Token } from "../Token";

describe("<Token/>", () => {
  let container: HTMLElement;
  const balance = "0";
  const totalBalance = "33.33";
  const price = 99.99;
  const decimals = 9;
  const onChangeExpanded = jest.fn();
  const onAddTokenToWallet = jest.fn();

  beforeEach(() => {
    const component = render(
      <Token
        expanded={false}
        onChangeExpanded={onChangeExpanded}
        onAddTokenToWallet={onAddTokenToWallet}
        decimals={decimals}
        symbol={"OHM"}
        address={""}
        icon={"33T"}
        balance={balance}
        price={price}
        totalBalance={totalBalance}
      />,
    );
    container = component.container;
  });

  it("Should match previous snapshot", () => {
    expect(container).toMatchSnapshot();
  });

  it("shows the total balance and value in USD", () => {
    const sigFigs = 5;
    const precision = 2;
    const balanceValueUSD = parseFloat(totalBalance) * price;

    const tokenBalanceEl = getByTestId(container, "balance-token");
    const usdBalanceEl = getByTestId(container, "balance-usd");

    expect(tokenBalanceEl).toBeInTheDocument();
    expect(usdBalanceEl).toBeInTheDocument();

    expect(tokenBalanceEl.textContent).toBe(totalBalance.substring(0, sigFigs));
    expect(usdBalanceEl.textContent).toBe(formatCurrency(balanceValueUSD, precision));
  });

  it("call onChangeExpanded when clicked", () => {
    const mainButton = screen.getByRole("button", { name: /ohm/i });
    expect(mainButton).toBeInTheDocument();

    userEvent.click(mainButton);

    expect(onChangeExpanded).toHaveBeenCalled();
  });
});
