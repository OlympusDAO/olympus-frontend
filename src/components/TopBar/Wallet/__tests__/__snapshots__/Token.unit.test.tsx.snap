// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Token/> Should match previous snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAccordion-root WithStyles(ForwardRef(Accordion))-root-1 MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root-3"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content-4"
      >
        <div
          class="MuiBox-root MuiBox-root-6"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            focusable="false"
            style="font-size: 28px; margin-right: 8px;"
            viewBox="0 0 32 32"
          >
            <defs>
              <style>
                .token_33T_svg__cls-2{isolation:isolate}.token_33T_svg__cls-3{fill:#fff}
              </style>
            </defs>
            <path
              d="M3.552 14.198c0-6.873 5.574-12.447 12.447-12.447 6.874 0 12.448 5.574 12.448 12.447 0 6.873-5.574 12.446-12.448 12.446-6.873 0-12.447-5.573-12.447-12.446Z"
              style="fill: #708b96;"
            />
            <g
              id="token_33T_svg__layer4"
              transform="matrix(.032 0 0 .032 0 -1.748)"
            >
              <g
                class="token_33T_svg__cls-2"
                id="token_33T_svg__text20110"
              >
                <g
                  class="token_33T_svg__cls-2"
                >
                  <path
                    class="token_33T_svg__cls-3"
                    d="M317.71 642.85v40.21h-5.89v-40.21H299v-5.63h31.5v5.63Zm72.53 17.29a23.21 23.21 0 0 1-6.79 17 23.8 23.8 0 0 1-33.54 0 24.75 24.75 0 0 1 0-34.06 23.72 23.72 0 0 1 33.54.07 23.12 23.12 0 0 1 6.79 16.99Zm-41.1 0a18.13 18.13 0 0 0 5 13.12 16.82 16.82 0 0 0 12.55 5.19 16.59 16.59 0 0 0 12.48-5.19 19.54 19.54 0 0 0 0-26.24 16.59 16.59 0 0 0-12.48-5.19 16.82 16.82 0 0 0-12.56 5.17 18.13 18.13 0 0 0-4.99 13.14ZM451.31 662q0 9.6-5.69 15.3-6.4 6.78-16.84 6.78a23.94 23.94 0 1 1 0-47.88 25.56 25.56 0 0 1 9.54 1.79 19 19 0 0 1 7.36 5.06l-4.22 4.22a14.31 14.31 0 0 0-5.48-4 18.08 18.08 0 0 0-7.2-1.44A17.25 17.25 0 0 0 416.1 647a19.37 19.37 0 0 0 0 26.24 17.25 17.25 0 0 0 12.68 5.19 17.07 17.07 0 0 0 11.4-3.84 15.34 15.34 0 0 0 5.25-10.56h-16.65v-5.51H451a19 19 0 0 1 .31 3.48Zm44.56-24.78v5.63H475v14.54h18.82v5.5H475v14.54h20.87v5.63h-26.76v-45.84Zm34.19 5.63v40.21h-5.89v-40.21h-12.81v-5.63h31.5v5.63Zm35.46 40.21h-5.89v-45.84h5.89v19.66h22.92v-19.66h5.89v45.84h-5.89v-20.68h-22.92Zm75.42-45.84v5.63h-20.87v14.54h18.82v5.5h-18.82v14.54h20.87v5.63h-26.76v-45.84Zm24.45 45.84h-5.89v-45.84h15.62a14.28 14.28 0 0 1 10.15 3.94 13.17 13.17 0 0 1 1.09 18.12 13.79 13.79 0 0 1-7.84 4.7l-.13.19 12.87 18.63v.26h-7l-12.36-18.57h-6.53Zm0-40.21V659h9.48a8.55 8.55 0 0 0 6.08-2.34 7.64 7.64 0 0 0 2.5-5.79 8.11 8.11 0 0 0-8.2-8Z"
                  />
                </g>
              </g>
            </g>
            <g
              id="token_33T_svg__g93605"
              transform="matrix(.032 0 0 .032 0 -1.748)"
            >
              <path
                class="token_33T_svg__cls-3"
                d="m537.9 302.69.26 63.46 32.49.22c8.73-18.92 24.19-35.55 49.83-35.58s47 22.23 46.91 45.38c0 28.15-23.62 50.24-47.62 50.48-14.55.14-31.34-1.8-46.92-13.92l-34.49-.39.2 75.46h33.68C592 474.07 605.93 473 617.5 473c23 0 48.78 17.53 49 46.15s-21.58 49.45-45.26 49.22c-24.21-.23-42.51-14.94-48.58-34.45l-33.81-.35-.58 63.2 28.23.24.28-19.77c13.48 14.16 38.14 21.38 51.51 21.22 45.16-.54 76.89-41 77.14-78.53s-23-59.23-43.53-70.09c17.85-8.6 44.31-32.21 44-73.94s-38.46-75.08-76.5-75.12c-21.37 0-39.59 10.3-52.57 21.63l-.17-19.46Z"
                id="token_33T_svg__path93601"
              />
              <path
                class="token_33T_svg__cls-3"
                d="m325.64 302.78.26 63.46 32.49.22c8.73-18.92 24.2-35.56 49.84-35.59s47 22.24 46.91 45.39c-.05 28.14-23.62 50.24-47.62 50.47-14.56.14-31.34-1.79-46.92-13.91l-34.49-.39.2 75.46H360c19.73-13.74 33.69-14.84 45.26-14.84 23 0 48.77 17.53 49 46.15s-21.62 49.45-45.26 49.22c-24.21-.23-42.51-14.94-48.58-34.45l-33.8-.35-.58 63.2 28.23.23.28-19.76c13.45 14.16 38.08 21.38 51.45 21.22 45.15-.55 76.89-41 77.14-78.53s-23-59.23-43.53-70.1c17.84-8.59 44.31-32.2 44-73.93s-38.46-75.08-76.49-75.12c-21.37 0-39.59 10.29-52.57 21.63l-.13-19.46Z"
                id="token_33T_svg__path93603"
              />
            </g>
            <path
              d="M15.983 28.253c7.705 0 13.95-6.246 13.95-13.951a13.93 13.93 0 0 0-.78-4.617V5.138A15.93 15.93 0 0 1 32 14.251c0 8.834-7.164 15.998-15.998 15.998S0 23.085 0 14.251A15.943 15.943 0 0 1 2.88 5.09v4.409a13.923 13.923 0 0 0-.848 4.803c0 7.705 6.246 13.951 13.951 13.951Z"
              style="fill: #d1ceaa;"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            OHM
          </p>
        </div>
        <div
          class="MuiBox-root MuiBox-root-8"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
            data-testid="balance-token"
            style="font-weight: 600;"
          >
            33.33
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            data-testid="balance-usd"
          >
            $3,332.67
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M10.4 6.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm0 10.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm-1.75-7a1.75 1.75 0 1 0 3.5 0 1.75 1.75 0 0 0-3.5 0Z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
              style="margin: auto; padding: 8px 0px;"
            >
              <div
                class="MuiBox-root MuiBox-root-10"
                style="gap: 8px;"
              >
                <div
                  class="MuiBox-root MuiBox-root-11 ohm-pairs"
                  style="width: 100%;"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-disableElevation MuiButton-fullWidth"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                      >
                        Add to Wallet
                      </p>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
